flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/executor/TestExecutorExtension.java

```diff
@@ -47,8 +58,26 @@ public T getExecutor() {
 
     @Override
     public void afterAll(ExtensionContext context) throws Exception {
+        gracefulShutdown(executorService, LOG);
+    }
+
+    static void gracefulShutdown(@Nullable ExecutorService executorService, Logger logger) {
```

Comment:
Good point. The rational behind this change was trying to make sure that any resources (that are handled by the threads) are actually cleaned up. What do you think about calling `shutdownNow` and failing with an `AssertionError` if there are still `Runnables` that have been processed by the executor? That way, we force everyone to handle hanging threads within the test implementation :thinking:

Reply:
> failing with an AssertionError if there are still Runnables that have been processed by the executor

That'd be quite annoying when using a scheduler executor service, no?

Reply:
That's true. But we could make this an opt-out feature. :thinking: 

But more generally asking: Don't you think that out-standing tasks that are left hanging might reveal some misbehavior in the test? I'm not passionate about that one and would be ok with using `shutdownNow()`. My only intention was to reveal issues in the code base that are missed right now.

Reply:
I can see where you're coming from, but think that for most tests this probably isn't relevant.
If an executor is used for the setup of some component that requires it then the test usually doesn't know what is even being scheduled.

This should only matter in tests that really want to drill down into the scheduling of operations aka "if I do _this_ then exactly _that and nothing else_ is being scheduled", and those are rather rare I think. The FutureUtilsTest is an example where this could be interesting.
An opt-in flag may be more appropriate.